src/ripple/app/tx/impl/TxQImpl.cpp

```diff
@@ -0,0 +1,969 @@
+        // grow without bound.
+        txnsExpected_ = feeLevels.size();
+    }
+
+    if (feeLevels.empty())
+    {
+        escalationMultiplier_ = minimumMultiplier_;
+    }
+    else
+    {
+        // In the case of an odd number of elements, this
+        // evaluates to the middle element; for an even
+        // number of elements, it will add the two elements
+        // on either side of the "middle" and average them.
+        escalationMultiplier_ = (feeLevels[size / 2] +
+            feeLevels[(size - 1) / 2] + 1) / 2;
+        escalationMultiplier_ = std::max(escalationMultiplier_,
+            minimumMultiplier_);
+    }
+    JLOG(j_.debug) << "Expected transactions updated to " <<
+        txnsExpected_ << " and multiplier updated to " <<
+        escalationMultiplier_;
+
+    return size;
+}
+
+std::uint64_t
+FeeMetrics::scaleFeeLevel(OpenView const& view) const
+{
+    auto fee = baseLevel;
+
+    // Transactions in the open ledger so far
+    auto const current = view.txCount();
+
+    std::lock_guard <std::mutex> sl(lock_);
+
+    // Target number of transactions allowed
+    auto const target = txnsExpected_;
+
+    // Once the open ledger bypasses the target,
+    // escalate the fee quickly.
+    if (current > target)
+    {
+        // Compute escalated fee level
+        fee = mulDivNoThrow(fee, current * current *
+            escalationMultiplier_,
+            target * target);
+    }
+
+    return fee;
+}
+
+//////////////////////////////////////////////////////////////////////////
+
+class TxQImpl : public TxQ
+{
+public:
+
+    TxQImpl(Setup const& setup,
+        beast::Journal j)
+        : setup_(setup)
+        , j_(j)
+        , feeMetrics_(setup.standAlone, j)
+    {
+    }
+
+    ~TxQImpl() override
+    {
+        byFee_.clear();
+    }
+
+    std::pair<TER, bool>
+    apply(Application& app, OpenView& view,
+        std::shared_ptr<STTx const> const& tx,
+        ApplyFlags flags, beast::Journal j) override;
+
+    bool
+    accept(Application& app, OpenView& view,
+        ApplyFlags flags = tapNONE) override;
+
+    void
+    processValidatedLedger(Application& app,
+        OpenView const& view, bool timeLeap,
+            ApplyFlags flags = tapNONE) override;
+
+    /** Used by tests only.
+    */
+    std::size_t
+    setMinimumTx(int m) override
+    {
+        return feeMetrics_.setMinimumTx(m);
+    }
+
+    struct Metrics
+    getMetrics(OpenView const& view) const override;
+
+    Json::Value
+    doRPC(Application& app) const override;
+
+    XRPAmount
+    openLedgerFee(OpenView const& view) const override;
+
+private:
+
+    using FeeHook = boost::intrusive::member_hook
+        <CandidateTxn, boost::intrusive::set_member_hook<>,
+        &CandidateTxn::byFeeListHook_>;
+
+    using FeeMultiSet = boost::intrusive::multiset
+        < CandidateTxn, FeeHook,
+        boost::intrusive::compare <GreaterFee> >;
+
+    using AccountSet = std::map < AccountID, TxQAccount >;
```

Comment:
Done. I'm leaving `FeeHook` and `FeeMultiSet`, thought, because they're more complicated.